### PR TITLE
[clang][Sema] Warn on self move for inlined static cast

### PR DESCRIPTION
**Pull request by MaxEW707**
There are code bases that inline `std::move` manually via `static_cast`.
Treat a static cast to an xvalue as an inlined `std::move` call and warn on a self move.

### CODE REVIEW

File: clang/lib/Sema/SemaChecking.cpp [hunk 1]
```diff
@@ -18843,17 +18843,26 @@ void Sema::DiagnoseSelfMove(const Expr *LHSExpr, const Expr *RHSExpr,
   LHSExpr = LHSExpr->IgnoreParenImpCasts();
   RHSExpr = RHSExpr->IgnoreParenImpCasts();
 
-  // Check for a call expression
+  // Check for a call expression or static_cast expression
   const CallExpr *CE = dyn_cast<CallExpr>(RHSExpr);
-  if (!CE || CE->getNumArgs() != 1)
+  const CXXStaticCastExpr *CXXSCE = dyn_cast<CXXStaticCastExpr>(RHSExpr);
```
Comment (tbaederr) [inline clang/lib/Sema/SemaChecking.cpp:18848]:
```suggestion
  const auto *CXXSCE = dyn_cast<CXXStaticCastExpr>(RHSExpr);
```

**Comment by MaxEW707** [comment]
Ping

File: clang/test/SemaCXX/warn-self-move.cpp [hunk 2]
```diff
@@ -35,11 +41,14 @@ class field_test {
   int x;
   field_test(field_test&& other) {
     x = std::move(x);  // expected-warning{{explicitly moving}}
+    x = static_cast<int&&>(x);  // expected-warning{{explicitly moving}}
```
Comment (shafik) [inline clang/test/SemaCXX/warn-self-move.cpp:44]: Can you add some tests where we should not get diagnostics for example like the one right below but w/ `static_cast`

**Review by shafik — APPROVED** [review]
LGTM, any objections @tbaederr

File: clang/lib/Sema/SemaChecking.cpp [hunk 3]
```diff
@@ -18843,17 +18843,26 @@ void Sema::DiagnoseSelfMove(const Expr *LHSExpr, const Expr *RHSExpr,
...
+  // Check for a static_cast<T&&>(..) to an xvalue which we can treat as an
+  // inlined std::move
+  if (CXXSCE && !CXXSCE->isXValue())
+    return;
+
+  // Get argument from std::move or static_cast
+  if (CE)
+    RHSExpr = CE->getArg(0);
+  else
+    RHSExpr = CXXSCE->getSubExpr();
```
Comment (tbaederr) [inline clang/lib/Sema/SemaChecking.cpp:18865]:
```suggestion
  if (const auto *CE = dyn_cast<CallExpr>(RHSExpr);
      CE && CE->getNumArgs() == 1 && CE->isCallToStdMove())
    RHSExpr = CE->getArg(0);
  else if (const auto *CXXSCE = dyn_cast<CXXStaticCastExpr>(RHSExpr);
      CXXSCE && CXXSCE->isXValue())
    RHSExpr = CXXSCE->getSubExpr();
  else
    return;
```

I think that would be shorter and cleaner. (This replaces the few lines above as well but I can't add a suggestion like that in Github).

**Review by tbaederr — APPROVED** [review]
Much better, thanks.

**Comment by MaxEW707** [comment]
Thanks all for the review :).

I will need one of you to commit on my behalf since I do not have write access.

**Pull request merged**